### PR TITLE
Grid integration double loop part 1

### PR DESCRIPTION
**Pull request by xiangao1**
## Summary/Motivation:
Open this PR to replace #638. 

## Changes proposed in this PR:
- Copied necessary files from #638 
- The new PR was needed because #638 originated from a branch on @Robbybp's fork which included some experimental changes to Caprese, which should not be included in this PR.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bknueven** [comment]
Is this waiting on / require #652?

**Comment by xiangao1** [comment]
> Is this waiting on / require #652?

Yes, @bknueven . I just talked to @adowling2 about this. I agreed to review #652 and get it merged, so I can have the integration tests ready for this PR.

File: idaes/apps/grid_integration/tests/test_bidder.py
Comment (lbianchi-lbl) [inline idaes/apps/grid_integration/tests/test_bidder.py:2]:
```suggestion
pytest.importorskip("prescient", reason="prescient is a requirement to run grid integration tests")

import pyomo.environ as pyo
```

File: idaes/apps/grid_integration/tests/test_coordinator.py
Comment (lbianchi-lbl) [inline idaes/apps/grid_integration/tests/test_coordinator.py:1]:
```suggestion
import pytest

pytest.importorskip("prescient", reason="prescient is a requirement to run grid integration tests")

```

File: idaes/apps/grid_integration/tests/test_tracker.py
Comment (lbianchi-lbl) [inline idaes/apps/grid_integration/tests/test_tracker.py:1]:
```suggestion
import pytest

pytest.importorskip("prescient", reason="prescient is a requirement to run grid integration tests")

```

**Review by lbianchi-lbl — CHANGES_REQUESTED** [review]
@xiangao1 @bknueven I'm working to fix the remaining conflicts in #652 so that it can be merged.

In the meantime, since Prescient is going to be an optional dependency for IDAES, we should make sure that the tests that require Prescient are skipped without errors/failures when Prescient it is not available. One way to do this is to use `pytest.importorskip()`; I've added how it can be used to the test files in the PR as review comments.

**Comment by lbianchi-lbl** [comment]
@xiangao1 There are still failures due to the optional `prescient` dependency not being skipped properly:

```log
_________ ERROR collecting apps/grid_integration/tests/test_bidder.py __________
ImportError while importing test module '/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/share/miniconda/envs/idaes-env/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py:3: in <module>
    from idaes.apps.grid_integration.bidder import Bidder
/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/__init__.py:3: in <module>
    from .coordinator import DoubleLoopCoordinator
/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/coordinator.py:1: in <module>
    import prescient.plugins as pplugins
E   ModuleNotFoundError: No module named 'prescient'
```

The reason #791 could be merged despite of this is that there are no branch protection rules that prevent merges until all required CI checks pass. Given this, I suggest that, throughout the review process for this PR, you push directly to the head branch (`IDAES:grid-int-double-loop`) rather than creating "second-order PRs" targeting it.

**Comment by xiangao1** [comment]
> @xiangao1 There are still failures due to the optional `prescient` dependency not being skipped properly:
> 
> ```
> _________ ERROR collecting apps/grid_integration/tests/test_bidder.py __________
> ImportError while importing test module '/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/importlib/__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py:3: in <module>
>     from idaes.apps.grid_integration.bidder import Bidder
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/__init__.py:3: in <module>
>     from .coordinator import DoubleLoopCoordinator
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/coordinator.py:1: in <module>
>     import prescient.plugins as pplugins
> E   ModuleNotFoundError: No module named 'prescient'
> ```
> 
> The reason #791 could be merged despite of this is that there are no branch protection rules that prevent merges until all required CI checks pass. Given this, I suggest that, throughout the review process for this PR, you push directly to the head branch (`IDAES:grid-int-double-loop`) rather than creating "second-order PRs" targeting it.

Understood.

**Comment by xiangao1** [comment]
> @xiangao1 There are still failures due to the optional `prescient` dependency not being skipped properly:
> 
> ```
> _________ ERROR collecting apps/grid_integration/tests/test_bidder.py __________
> ImportError while importing test module '/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/importlib/__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py:3: in <module>
>     from idaes.apps.grid_integration.bidder import Bidder
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/__init__.py:3: in <module>
>     from .coordinator import DoubleLoopCoordinator
> /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/coordinator.py:1: in <module>
>     import prescient.plugins as pplugins
> E   ModuleNotFoundError: No module named 'prescient'
> ```
> 
> The reason #791 could be merged despite of this is that there are no branch protection rules that prevent merges until all required CI checks pass. Given this, I suggest that, throughout the review process for this PR, you push directly to the head branch (`IDAES:grid-int-double-loop`) rather than creating "second-order PRs" targeting it.

@lbianchi-lbl How can I know if I fixed this on my local machine? Thanks!

**Comment by lbianchi-lbl** [comment]
> > @xiangao1 There are still failures due to the optional `prescient` dependency not being skipped properly:
> > ```
> > _________ ERROR collecting apps/grid_integration/tests/test_bidder.py __________
> > ImportError while importing test module '/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py'.
> > Hint: make sure your test modules/packages have valid Python names.
> > Traceback:
> > /usr/share/miniconda/envs/idaes-env/lib/python3.8/importlib/__init__.py:127: in import_module
> >     return _bootstrap._gcd_import(name[level:], package, level)
> > /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/tests/test_bidder.py:3: in <module>
> >     from idaes.apps.grid_integration.bidder import Bidder
> > /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/__init__.py:3: in <module>
> >     from .coordinator import DoubleLoopCoordinator
> > /usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/coordinator.py:1: in <module>
> >     import prescient.plugins as pplugins
> > E   ModuleNotFoundError: No module named 'prescient'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The reason #791 could be merged despite of this is that there are no branch protection rules that prevent merges until all required CI checks pass. Given this, I suggest that, throughout the review process for this PR, you push directly to the head branch (`IDAES:grid-int-double-loop`) rather than creating "second-order PRs" targeting it.
> 
> @lbianchi-lbl How can I know if I fixed this on my local machine? Thanks!

One way to do that is to uninstall Prescient (`pip uninstall gridx-prescient`), run the tests, apply `pytest.importorskip()` as needed, and reinstall Prescient when done.

**Comment by adowling2** [comment]
Another way would be to make yourself a new conda env for testing without Prescient.

**Comment by lbianchi-lbl** [comment]
@xiangao1 There are two failures in the non-integration tests that should be relatively straightforward to fix:

- [ ] `test_headers`: run the following command from the root of the repository: `addheader -c addheader.yml idaes/apps/grid_integration/`
- [ ] `test_imports`: This seems to be due to `'idaes.apps.grid_integration.examples.thermal_generator_prescient_plugin`. I'm not sure what's the plan with the examples but in the meantime you could wrap problematic imports in a `try/except ImportError` block

File: idaes/apps/grid_integration/examples/thermal_generator_prescient_plugin.py
Comment (lbianchi-lbl) [inline idaes/apps/grid_integration/examples/thermal_generator_prescient_plugin.py:8]: If this refers to the `thermal_generator.py` file in the same directory, it should work using a relative import:

```suggestion
from .thermal_generator import ThermalGenerator
```

Comment (xiangao1) [reply idaes/apps/grid_integration/examples/thermal_generator_prescient_plugin.py:8]: Thanks! I fixed it and will push later.

**Comment by lbianchi-lbl** [comment]
Some useful commands if you want to generate a test coverage report locally:

```sh
pytest --cov --cov-report html:.cov-report -k grid_integration  # -k grid_integration selects only the tests within the grid_integration directory
firefox .cov-report/index.html  # or open this file with any other web browser
```

image

image

**Review by lbianchi-lbl — DISMISSED** [review]

**Comment by lbianchi-lbl** [comment]
@bknueven @dguittet @adowling2: @xiangao1 and I managed to get all the code infrastructure to work correctly, and now all tests pass. Feel free to review it, and then we should be able to merge this PR as soon as it's approved.

File: idaes/apps/grid_integration/examples/thermal_generator.py
Comment (adowling2) [inline idaes/apps/grid_integration/examples/thermal_generator.py:450]: Is it possible to add a simple test to remove these codecov warnings?

**Review by adowling2 — COMMENTED** [review]
Review of about 1/2 the files.

File: idaes/apps/grid_integration/bidder.py
Comment (bknueven) [inline idaes/apps/grid_integration/bidder.py:85]: Would it be better to have an ABC for the bidding_model_object, e.g. `AbstractBiddingModel`. Then we could just check `isinstance(self.bidding_model_object, AbstractBiddingModel)`. I see this is an issue further in the code -- might best to be just enforced through ABCs.

As-it-is, this code defines the expected interface. This might be better:
```python
class AbstractBiddingModel(ABC):
    @abstractmethod
    def populate_model(self, b):
        '''
        documentation about what the user should implement
        '''
        pass

    @abstractmethod
    def update_model(self, b, implemented_shut_down, implemented_start_up, implemented_power_output):
        '''
        documentation about what should be implemented
        '''
        pass

    @property
    @abstractmethod
    def power_output(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def total_cost(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def generator(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def pmin(self):
        '''
        documentation about this property
        '''
        pass

    @property
    @abstractmethod
    def default_bids(self):
        '''
        documentation about this property
        '''
        pass

```

File: idaes/apps/grid_integration/bidder.py
Comment (bknueven) [inline idaes/apps/grid_integration/bidder.py:28]: We would ideally have documentation and the function signature for these abstract methods, so those inheriting from them know what to expect.

File: idaes/apps/grid_integration/bidder.py
Comment (bknueven) [inline idaes/apps/grid_integration/bidder.py:47]: `abstractproperty` was depreciated in Python 3.3. https://docs.python.org/3/library/abc.html#abc.abstractproperty

File: idaes/apps/grid_integration/bidder.py
Comment (bknueven) [inline idaes/apps/grid_integration/bidder.py:531]: Any reason to use a `ConstraintList` here as opposed to building a normal indexed constraint? We don't seem to mutate them.

If this is a matter of the `time_index` depending on `k[0]`, that could be handled other ways, e.g.,
```python
self.model.bidding_constraints = pyo.Constraint(self.model.SCENARIOS, self.model.SCENARIOS, pyo.Any)
```
or, keeping the `scenario_comb`:
```python
self.model.bidding_constraints = pyo.Constraint(scenario_comb, pyo.Any)
```
though obviously it would be preferable to use an explicit time index for the last component in either case. 
Note that you can simply skip the case where the scenarios are the same using the first version.

File: idaes/apps/grid_integration/bidder.py
Comment (bknueven) [inline idaes/apps/grid_integration/bidder.py:704]: There's some code in Egret (which this implicitly depends on) which does this as well. https://github.com/grid-parity-exchange/Egret/blob/42eb712299cb32ee1e65e97069ee9b4320f49f3d/egret/model_library/transmission/tx_utils.py#L611.

You might try using that instead of implementing your own. This file demonstrates its use: https://github.com/grid-parity-exchange/Egret/blob/main/egret/model_library/transmission/tests/test_tx_utils.py

File: idaes/apps/grid_integration/coordinator.py
Comment (bknueven) [inline idaes/apps/grid_integration/coordinator.py:470]: I worry a bit about using `zip` on `component_objects` like this. I suppose it'll be okay if they're literally the same Pyomo model...

@jsiirola thoughts?

**Review by bknueven — COMMENTED** [review]
I have a few comments as well -- haven't yet gotten to review every file.

File: idaes/apps/grid_integration/bidder.py
Comment (xiangao1) [reply idaes/apps/grid_integration/bidder.py:85]: @bknueven Thanks, Ben! This makes sense. It seems to be more changes than other suggestions. Should I open another PR to address this?

File: idaes/apps/grid_integration/tests/integration_test_plugin.py
Comment (bknueven) [inline idaes/apps/grid_integration/tests/integration_test_plugin.py:77]: Any reason not to just import these from the example for now instead of replicating all the code here?

File: idaes/apps/grid_integration/tests/test_integration.py
Comment (bknueven) [inline idaes/apps/grid_integration/tests/test_integration.py:132]: Ideally we would check the output in this file, `tracker_detail.csv`, and `tracking_model_detail.csv`, but maybe for later.

File: idaes/apps/grid_integration/bidder.py
Comment (bknueven) [inline idaes/apps/grid_integration/bidder.py:267]: This also doesn't seem like it needs to be a `ConstraintList`.

**Review by bknueven — COMMENTED** [review]
Lots of small comments, none really critical.

Big picture question: how is the time configuration handled? Do we expect users to pass it into the coordinator/bidder/tracker objects? Minimally we'd get the time configuration out of Prescient and at least validate it against what the user passed in -- best practice would be to just read it out of the Prescient configuration.

File: idaes/apps/grid_integration/bidder.py
Comment (bknueven) [reply idaes/apps/grid_integration/bidder.py:158]: I don't think this is critical -- if you want to address it in a follow on PR I think that is okay.

File: idaes/apps/grid_integration/tests/test_integration.py
Comment (xiangao1) [reply idaes/apps/grid_integration/tests/test_integration.py:132]: I will take a note on this and address it later.

File: idaes/apps/grid_integration/coordinator.py
Comment (bknueven) [inline idaes/apps/grid_integration/coordinator.py:342]:
```python
        gen_dict["p_min"] = p_min_scalar
```
or
```python
        gen_dict["p_min"] = {'data_type':'time_series', 'values': list_of_pmin}
```
`p_max` can be set similarly.


To fix a generator on
```python
        gen_dict["fixed_commitment"] = 1
```
also fix it to schedule:
```python
        gen_dict["fixed_commitment"] = {'data_type':'time_series', 'values': list_of_0_1_commitments}
        gen_dict["min_up_time"] = 0
        gen_dict["min_down_time"] = 0
```

Could also be set in the real-time updater.

File: idaes/apps/grid_integration/bidder.py
Comment (adowling2) [reply idaes/apps/grid_integration/bidder.py:158]: If you go with a follow-up PR, we should create an issue with a checklist of everything we want in the follow-up PR(s).

File: idaes/apps/grid_integration/examples/thermal_generator.py
Comment (adowling2) [reply idaes/apps/grid_integration/examples/thermal_generator.py:447]: If this formulation is from our paper or literature, I am okay with just adding a comment to cite the correct reference and appropriate equations. IMO, for a test problem, documenting the source is more important than fine-tuning the model (provided it works).

File: idaes/apps/grid_integration/forecaster.py
Comment (adowling2) [inline idaes/apps/grid_integration/forecaster.py:39]: Perhaps add a block comment here explaining this is a simple test. Or does it make sense to move this to a specific test file? I think the latter would resolve the Codecov warning.

File: idaes/apps/grid_integration/tests/integration_test_plugin.py
Comment (adowling2) [reply idaes/apps/grid_integration/tests/integration_test_plugin.py:77]: Yes, let's have a follow-up conversation about this. Please create an issue to track this (and other) follow-up discussions. You can link specifically to this comment thread.

File: idaes/apps/grid_integration/tests/util.py
Comment (adowling2) [inline idaes/apps/grid_integration/tests/util.py:21]: Does this model come from literature (is it in one of our papers)? If so, just add a quick reference here in the comment block.

File: idaes/apps/grid_integration/tests/util.py
Comment (adowling2) [inline idaes/apps/grid_integration/tests/util.py:32]: Update doc string to match function inputs. Add units to docstring.

File: idaes/apps/grid_integration/tests/util.py
Comment (adowling2) [inline idaes/apps/grid_integration/tests/util.py:24]: Add comment with units

**Review by adowling2 — COMMENTED** [review]
This is getting really close. Here are the outstanding issues I see:
- [x] Add simple unit or regression tests for self-schedule. This will take care of many of the codecov warnings.
- [ ] Open a new issue to track double loop enhancements if we do not have one already. Add topics raised here to the issue and link as appropriate. For example:
   - [ ] Follow-up discussion with @bknueven on improving the abstraction
   - [ ] Correct usage of IDAES logger
   - [ ] Add Rankine cycle example and appropriate tests. Resolve Codecov warning for multiperiod.
   - [ ] Anything else??

**Comment by xiangao1** [comment]
Hi @adowling2, @bknueven, @lbianchi-lbl, I have addressed most of you comments. Thank you for your help with the PR. Given @lbianchi-lbl is on vacation next week, we are targeting to get this merged tomorrow.

Ben would you mind taking a look at the following commits:
https://github.com/IDAES/idaes-pse/pull/787/commits/3e5fedfd1c9548dd929adc2a2fc43e40ebfba176
https://github.com/IDAES/idaes-pse/pull/787/commits/66630cb59af36fc2c2b39e3ccb488ebc038686b7

The two commits are mainly for the generalization of bid structure from self-scheduler and bidder (what we have been discussing on slack).

**Comment by bknueven** [comment]
We should be sure any time we're updating `p_min` that `shutdown_capacity` and `startup_capacity` are appropriately modified to avoid Egret's somewhat picky validators. (In particular, `shutdown_capacity >= p_min` and `startup_capacity >= p_min`).

I had a question [here](https://github.com/IDAES/idaes-pse/pull/787/files/be68ea3d1a29550a5c3885e1915600a75357c665#r848868823) I really would like @jsiirola to weigh in on.

File: idaes/apps/grid_integration/coordinator.py
Comment (jsiirola) [inline idaes/apps/grid_integration/coordinator.py:86]: What is the value of these 7 1-line functions?  Why don't you just inline them here?

File: idaes/apps/grid_integration/coordinator.py
Comment (adowling2) [reply idaes/apps/grid_integration/coordinator.py:514]: @jsiirola Thoughts?

Comment (jsiirola) [reply idaes/apps/grid_integration/coordinator.py:514]: This is really asking for trouble.  You are counting on the declaration order and all sets to match between the two models.  It might be sufficient to set `sort=SortComponents.alphabetizeComponentAndIndex` when calling `component_objects`.  You should still verify that the two component names match, though...

File: idaes/apps/grid_integration/coordinator.py
Comment (bknueven) [inline idaes/apps/grid_integration/coordinator.py:517]: @jsiirola what about something like this instead?

```suggestion
            for proj_tracker_obj in self.projection_tracker.model.component_objects(obj):
                tracker_obj = self.projection_tracker.model.find_component(proj_tracker_obj.name)
                for idx in proj_tracker_obj.index_set():
                    if pyo.value(proj_tracker_obj[idx]) != pyo.value(tracker_obj[idx]):
                        proj_tracker_obj[idx] = round(pyo.value(tracker_obj[idx]), 4)
```

**Comment by adowling2** [comment]
From my perspective, this PR is ready to merge once we:

- [ ] Address @bknueven's comment here. My interpretation is that adding a small epsilon to the values would ensure Egret's validators are satisfied.

> We should be sure any time we're updating `p_min` that `shutdown_capacity` and `startup_capacity` are appropriately modified to avoid Egret's somewhat picky validators. (In particular, `shutdown_capacity >= p_min` and `startup_capacity >= p_min`).
> 
> I had a question [here](https://github.com/IDAES/idaes-pse/pull/787/files/be68ea3d1a29550a5c3885e1915600a75357c665#r848868823) I really would like @jsiirola to weigh in on.

- [ ] Converge on @bknueven's suggestion to not use `zip`:
https://github.com/IDAES/idaes-pse/pull/787#discussion_r851318845

File: idaes/apps/grid_integration/coordinator.py
Comment (jsiirola) [reply idaes/apps/grid_integration/coordinator.py:517]: That would work, too.  I think I would go with the `find_component()` route if you expect that one model may have more objects than the other.  Otherwise, I would expect that the `sort=` would be more efficient.  We should note that the `find_component` route will be very slow through Pyomo 6.4.  Starting in 6.5 thinkgs will be faster (but I would expect still slower than the `sort=` route).

If we go the `sort=` route, I would also recommend using `zip_longest` and not `zip` (zip will quit when the first list is exhausted, so if the models are different sizes, we wouldn't see an error)

**Comment by bknueven** [comment]
> From my perspective, this PR is ready to merge once we:
> 
>     * [ ]  Address @bknueven's comment here. My interpretation is that adding a small epsilon to the values would ensure Egret's validators are satisfied.
> 
> 
> > We should be sure any time we're updating `p_min` that `shutdown_capacity` and `startup_capacity` are appropriately modified to avoid Egret's somewhat picky validators. (In particular, `shutdown_capacity >= p_min` and `startup_capacity >= p_min`).

It is sufficient just to set `shutdown_capacity` and `startup_capacity` to the `p_min`.

File: idaes/apps/grid_integration/coordinator.py
Comment (xiangao1) [reply idaes/apps/grid_integration/coordinator.py:86]: These functions don't return anything. Should I replace these with the actual implementation of the functions?

File: idaes/apps/grid_integration/coordinator.py
Comment (bknueven) [reply idaes/apps/grid_integration/coordinator.py:517]: > If we go the `sort=` route, I would also recommend using `zip_longest` and not `zip` (zip will quit when the first list is exhausted, so if the models are different sizes, we wouldn't see an error), and we should absolutely compare the component names to make sure that things are lining up correctly.

I like this. I suspect there's no reason to think the models will diverge, but at least this way we will get a reasonable error if it happens.

File: idaes/apps/grid_integration/coordinator.py
Comment (bknueven) [reply idaes/apps/grid_integration/coordinator.py:86]: > These functions don't return anything. Should I replace these with the actual implementation of the functions?

Yeah, I think that is cleaner as well, but probably not critical for merging.

Comment (xiangao1) [reply idaes/apps/grid_integration/coordinator.py:86]: Thanks!

File: idaes/apps/grid_integration/coordinator.py
Comment (xiangao1) [reply idaes/apps/grid_integration/coordinator.py]: I just pushed to address this! Could you take a look at it @bknueven ? Thanks!

Comment (bknueven) [reply idaes/apps/grid_integration/coordinator.py]: This looks good!

**Comment by xiangao1** [comment]
> We should be sure any time we're updating `p_min` that `shutdown_capacity` and `startup_capacity` are appropriately modified to avoid Egret's somewhat picky validators. (In particular, `shutdown_capacity >= p_min` and `startup_capacity >= p_min`).
> 
> I had a question [here](https://github.com/IDAES/idaes-pse/pull/787/files/be68ea3d1a29550a5c3885e1915600a75357c665#r848868823) I really would like @jsiirola to weigh in on.

@bknueven Could you look at this commit? https://github.com/IDAES/idaes-pse/pull/787/commits/a7af5f5dc438f2fc271aac2e7f4e49be4ffef964

Thanks!

**Review by adowling2 — APPROVED** [review]

**Review by bknueven — DISMISSED** [review]

**Review by jsiirola — APPROVED** [review]

**Comment by xiangao1** [comment]
@bknueven @adowling2 @jsiirola, I have fixed the test that failed.

**Review by bknueven — APPROVED** [review]

**Review by adowling2 — DISMISSED** [review]

**Comment by adowling2** [comment]
@lbianchi-lbl @ksbeattie This has been approved and is set to auto-merge. Once the integration tests are done, you should be ready to cut the alpha release.

**Comment by adowling2** [comment]
@xiangao1 One integration test failed:

```
         # check if bids are convex
        for t in bids:
            for gen in bids[t]:
                temp_curve = {
                    "data_type": "cost_curve",
                    "cost_curve_type": "piecewise",
                    "values": list(bids[t][gen].items()),
                }
>               tx_utils.validate_and_clean_cost_curve(
                    curve=temp_curve,
                    curve_type="cost_curve",
                    p_min=min(bids[t][gen].keys()),
                    p_max=max(bids[t][gen].keys()),
                    gen_name=gen,
                    t=t,
                )
E               NameError: name 'tx_utils' is not defined

/usr/share/miniconda/envs/idaes-env/lib/python3.8/site-packages/idaes/apps/grid_integration/bidder.py:843: NameError
```

**Review by bknueven — APPROVED** [review]

**Review by adowling2 — APPROVED** [review]

**Comment by adowling2** [comment]
Fingers crossed the tests pass.

**Pull request merged**
